server/routes/api/browser.py

```diff
@@ -164,10 +163,49 @@ def get_statvar_hierarchy(dcid):
     for svg_id, svg in svg_map.items():
         if svg_id in seen_sv:
             continue
+        svg["level"] = 0
         processed_svg_map[svg_id] = svg
         statvar_hierarchy_helper(svg_id, svg_map, processed_svg_map,
-                                 processed_sv, seen_sv)
+                                 processed_sv, seen_sv, 1)
+    for sv_id, sv in processed_sv.items():
+        parent_svg = processed_svg_map.get(sv["parent"])
+        sv["level"] = parent_svg["level"] + 1
     result = {}
     result["statVarGroups"] = processed_svg_map
     result["statVars"] = processed_sv
     return Response(json.dumps(result), 200, mimetype='application/json')
+
+
+@cache.memoize(timeout=3600 * 24)  # Cache for one day.
+def search_helper(token_string):
+    """gets the set of results that matches all the tokens in the token_string
+    
+    Args:
+        token_string: list of tokens as a string separated by ^
+    
+    Returns:
+        set of values that matches all the tokens in the token_string
+    """
+    tokens = token_string.split("^")
+    last_token_result = set(current_app.config['STAT_VAR_SEARCH_INDEX'].get(
+        tokens[-1], {}))
+    if len(tokens) == 1:
+        return last_token_result
+    prev_tokens = tokens[:(len(tokens) - 1)]
+    prev_tokens_string = "^".join(prev_tokens)
+    prev_intersection = search_helper(prev_tokens_string)
+    return last_token_result.intersection(prev_intersection)
+
+
+@bp.route('/search_statvar_hierarchy')
+@cache.cached(timeout=3600 * 24, query_string=True)
+def search_statvar_hierarchy():
+    """Gets the statvars and statvar groups that match the tokens in the query
+    """
+    query = request.args.get("query").lower()
+    query = query.replace(",", "")
+    tokens = query.split(" ")
+    sorted_tokens = sorted(tokens)
```

Comment:
sort is not really needed here?

Reply:
the sort was so that the same tokens in a different order can read from the cache instead of reprocessing the result

Reply:
I see. Some of the complexity here is to use the Flask cache. Since the operation here is to get intersection of in-memory sets without additional API call, you don't need to use the cache. Or to confirm, can profile how much time the operation takes, then we can decide?

Reply:
yeah, that makes sense

Reply:
profiled and it doesn't look like caching makes much of a difference. I removed the added complexity that was used for caching.